ssl/handshake_server.cc

```diff
@@ -264,32 +264,41 @@ static bool negotiate_version(SSL_HANDSHAKE *hs, uint8_t *out_alert,
   return true;
 }
 
-static UniquePtr<STACK_OF(SSL_CIPHER)> ssl_parse_client_cipher_list(
-    const SSL_CLIENT_HELLO *client_hello) {
+bool ssl_parse_client_cipher_list(
+    const SSL_CLIENT_HELLO *client_hello, UniquePtr<STACK_OF(SSL_CIPHER)> *ciphers_out) {
+  ciphers_out->reset();
+
   CBS cipher_suites;
   CBS_init(&cipher_suites, client_hello->cipher_suites,
            client_hello->cipher_suites_len);
 
+  // Cipher suites are encoded as 2-byte unsigned integers
+  if (CBS_len(&cipher_suites) % 2 != 0) {
+    return false;
```

Comment:
Should this push an error before returning? Previous error would have been a `SSL_R_ERROR_IN_RECEIVED_CIPHER_LIST` while trying to read a u16 from the CBS. Seems like this would mask malformed data.

Reply:
I guess this check was down in the TLS 1.3 chose cipher logic. I'm not sure what the protocol behavior should be here, and whether its the same between 1.2 & 1.3.

Reply:
I believe the prior behavior for odd length CBS would have been to [not set an error](https://github.com/aws/aws-lc/blob/main/ssl/s3_both.cc#L691) until two calls up the stack where [we'd set a `SSL_R_NO_SHARED_CIPHER` error and send a `SSL_AD_HANDSHAKE_FAILURE`](https://github.com/aws/aws-lc/blob/main/ssl/tls13_server.cc#L243-L245). This change currently sends a `SSL_AD_INTERNAL_ERROR` alert on parsing failure, which doesn't seem [per RFC 5246](https://datatracker.ietf.org/doc/html/rfc5246#section-7.2.2).

I'll change the alert sent to the client, and set the error to `SSL_R_ERROR_IN_RECEIVED_CIPHER_LIST` on parse failure as that seems most appropriate. I don't think setting the error actually impacts what goes on the wire, just how the error is reported locally.

Reply:
I think those changes proposed would be appropriate.